crates/lib/src/db/raw_def/v9.rs

```diff
@@ -315,8 +294,179 @@ pub struct RawReducerDefV9 {
+        ty: spacetimedb_sats::AlgebraicType,
+        custom_ordering: bool,
+    ) -> AlgebraicTypeRef {
+        let ty = self.module.typespace.add(ty);
+        let scope = scope.into_iter().collect();
+        let name = name.into();
+        self.module.types.push(RawTypeDefV9 {
+            name: RawScopedTypeNameV9 { name, scope },
+            ty,
+            custom_ordering,
+        });
+        // We don't add a `TypeId` to `self.type_map`, because there may not be a corresponding Rust type! e.g. if we are randomly generating types in proptests.
+        ty
+    }
+
+    /// Add a reducer to the in-progress module.
+    /// Accepts a `ProductType` of reducer arguments for convenience.
+    /// The `ProductType` need not be registered in the typespace.
+    ///
+    /// Importantly, if the reducer's first argument is a `ReducerContext`, that
+    /// information should not be provided to this method.
+    /// That is an implementation detail handled by the module bindings and can be ignored.
+    /// As far as the module definition is concerned, the reducer's arguments
+    /// start with the first non-`ReducerContext` argument.
+    ///
+    /// (It is impossible, with the current implementation of `ReducerContext`, to
+    /// have more than one `ReducerContext` argument, at least in Rust.
+    /// This is because `SpacetimeType` is not implemented for `ReducerContext`,
+    /// so it can never act like an ordinary argument.)
+    pub fn add_reducer(&mut self, name: impl Into<RawIdentifier>, params: spacetimedb_sats::ProductType) {
+        self.module.reducers.push(RawReducerDefV9 {
+            name: name.into(),
+            params,
+        });
+    }
+
+    /// Get the typespace of the module.
+    pub fn typespace(&self) -> &Typespace {
+        &self.module.typespace
+    }
+
+    /// Finish building, consuming the builder and returning the module.
+    /// The module should be validated before use.
+    pub fn finish(self) -> RawModuleDefV9 {
+        self.module
+    }
+}
+
+impl TypespaceBuilder for RawModuleDefV9Builder {
+    fn add(
+        &mut self,
+        typeid: TypeId,
+        name: Option<&'static str>,
+        make_ty: impl FnOnce(&mut Self) -> AlgebraicType,
+    ) -> AlgebraicType {
+        let r = match self.type_map.entry(typeid) {
+            btree_map::Entry::Occupied(o) => *o.get(),
+            btree_map::Entry::Vacant(v) => {
+                // Bind a fresh alias to the unit type.
+                let slot_ref = self.module.typespace.add(AlgebraicType::unit());
+                // Relate `typeid -> fresh alias`.
+                v.insert(slot_ref);
+
+                // Alias provided? Relate `name -> slot_ref`.
+                if let Some(name) = name {
+                    // Right now, we just split the name on patterns likely to split up module paths.
+                    // TODO(1.0): build namespacing directly into the bindings macros so that we don't need to do this.
```

Comment:
Note: this will not be an ABI break because we can rewrite the `TypespaceBuilder` trait when we are redoing the Rust binding macros to return v9. `TypespaceBuilder` lives *inside* the module, so it is not part of the ABI. It is part of the API surface touched by the macros, so this will be an *API* break, but it's one of those weird actually-private-despite-being-public APIs you get with macros.